executor/write.go

```diff
@@ -819,15 +844,32 @@ func (e *InsertValues) getRows(cols []*table.Column) (rows [][]types.Datum, err
 }
 
 func (e *InsertValues) getRow(cols []*table.Column, list []expression.Expression) ([]types.Datum, error) {
-	vals := make([]types.Datum, len(list))
-	for i, expr := range list {
+	length := len(e.Lists[0]) // It's length of elements coming from values(...) explicitly.
+	vals := make([]types.Datum, length)
+	for i, expr := range list[0:length] {
 		val, err := expr.Eval(nil)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
 		vals[i] = val
+	}
+	datums, err := e.fillRowData(cols, vals, false)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// For generated column, we should eval expr with a datum list
+	// which has same schema with e.Table. And, we can eval them one
+	// by one, because they can only refer generated columns occurring
+	// earilier in the table.
+	for i, expr := range list[length:] {
```

Comment:
In `executor/write.go`, all generated columns are always coms last. If you update your table **t** like this:
```sql
update t set a = 3, b = 2, e = 6;
```
the set list will be `a = 3, b = 2, e = 6, d = 3+2`.